File: Topological Sorting/1005번 - ACM Craft/1005.py
import sys
from collections import deque

def case():
    n, k = map(int, input().split())
    time = [0] + list(map(int, input().split()))
    res_time = time[:]
    graph = [[] for _ in range(n + 1)]
    indegree = [0] * (n + 1)

    for _ in range(k):
        x, y = map(int, input().split())
        graph[x].append(y)
        indegree[y] += 1

    w = int(input())

    Q = deque(i for i in range(1, n + 1) if indegree[i] == 0)

    while Q:
        cur = Q.popleft()
        for nxt in graph[cur]:
            res_time[nxt] = max(res_time[nxt], res_time[cur] + time[nxt])
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                Q.append(nxt)

    print(res_time[w])

def solve():
    t = int(input())
    for _ in range(t):
        case()
    
if __name__ == '__main__':
    solve()

File: Topological Sorting/1005번 - ACM Craft/1005.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int n, k, w;
int sp_tm[1001];
int res_tm[1001];
int indegree[1001];
vector<int> g[1001];

void input() {
    cin >> n >> k;
    for(int i = 1; i <= n; i++) {
        cin >> sp_tm[i];
        res_tm[i] = sp_tm[i];
    }
    for(int i = 0, x, y; i < k; i++) {
        cin >> x >> y;
        g[x].push_back(y);
        indegree[y]++;
    }
    cin >> w;
}

void solve() {
    queue<int> Q;
    for(int i = 1; i <= n; i++) {
        if(indegree[i] == 0) Q.push(i);
    }
    
    while(!Q.empty()) {
        int cur = Q.front(); Q.pop();
        
        for(int i = 0; i < g[cur].size(); i++) {
            int next = g[cur][i];
            res_tm[next] = max(res_tm[next], res_tm[cur] + sp_tm[next]);
            if(--indegree[next] == 0) Q.push(next);
        }
    }
    cout << res_tm[w] << "\n";
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int t;
    cin >> t;
    while(t--) {
        input();
        solve();
        for(int i = 1; i <= n; i++) g[i].erase(g[i].begin(), g[i].end());
    }
    
    return 0;
}